### PR TITLE
[PLATFORM-701]: Allow runtime config

### PR DESCRIPTION
**Pull request by cpiemontese**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-701

Allow `:endpoint` to be passed to `call` to override the base endpoint.

I also made the base `:endpoint` non-required, so that the user isn't forced to specify it if it doesn't fit their use case, but let's discuss it!

### CODE REVIEW

**Review by edoardocostantinidev — DISMISSED** [review]

**Review by zoten — APPROVED** [review]
The only gotcha is if the user does not set a global `endpoint` and then also `call`s without an `endpoint` opt, the runtime error may be unclear. Can we add some (even runtime) check on this case in your opinon? Or at least a README warning :P

**Comment by cpiemontese** [comment]
I was thinking about that! Should we return `{:error, "missing :endpoint"}` or would it be better to `raise`?

**Review by chrooti — COMMENTED** [review]
I feel like after the big OTP 25 migration the module level constant won't be used at all, so maybe we can make a call/4 function that takes endpoint as arg?

**Comment by chrooti** [comment]
> unluckily this would break backwards compatibility

Fair enough, I thought we didn't care much about that (since I hope everyone will have migrated in a few months from now).

>  I think call/3 and call/4 would collide on the call/3 definition.

Maybe we can make endpoint mandatory in `call/4` and having `call/3` read it from config (in order to deprecate it and give time to migrate to a hypotetical call/4 only version).

Which after all it's exactly what you proposed, only with an overload instead of another option.

**Comment by cpiemontese** [comment]
Why would OTP 25 make the static endpoint obsolete? Is it because we are migrating towards mix release?
Also: will this only affect `:endpoint` or are we moving towards runtime variables in general? I think I'm missing some context :sweat_smile: 

In any case, if this is the direction we are going in I like the idea of creating a `call/4`, it would allow us to provide a new `call` function which actually supports the new "flow" while keeping the older one as a deprecated option.

`call/3` could still depend on the base endpoint and raise if it's not present. I am a bit wary of using configurations because of [this](https://hexdocs.pm/elixir/1.13/library-guidelines.html#avoid-application-configuration).

**Comment by chrooti** [comment]
> Why would OTP 25 make the static endpoint obsolete? Is it because we are migrating towards mix release?

yeah

>  are we moving towards runtime variables in general?

we're moving to runtime variables where it makes sense, AKA when there is config that differs per-environment

**Comment by cpiemontese** [comment]
Ok I tried to fight a bit with Elixir to get the damn `call/4` to work along side `call/3` but Elixir won.

Are you ok with having the `:endpoint` in the options for now? Eventually, when people have migrated to OTP 25, we could create a breaking change, replace `call/3` with `call/4` and remove the base endpoint altogether

**Comment by edoardocostantinidev** [comment]
@cpiemontese I think it's a good compromise 👍🏻

**Comment by chrooti** [comment]
> Ok I tried to fight a bit with Elixir to get the damn call/4 to work along side call/3 but Elixir won.

the fix suggested by Christopher in the [thread](https://prima.slack.com/archives/C016E6R5ZLG/p1665743433630599?thread_ts=1665741250.583419&cid=C016E6R5ZLG) didn't work? 
sadge

**Comment by cpiemontese** [comment]
@chrooti it would work but it would make it so that `call/4` would have un-optional `opts` and you'll have to pass `[]` explicitly

It's not dramatically bad but also not really good :laughing:

**Comment by zoten** [comment]
I'm still convinced we can manage to make it work with `call/3` AND also have "automatic" endpoint resolution without having to override manually the endpoint at each call, e.g.

```
# lib/graphql.ex:108
endpoint = Keyword.get(opts, :endpoint, @endpoint)
```

could become

```
# lib/graphql.ex:108
endpoint = opts |> Keyword.get(:endpoint, @endpoint) |> BridgeEx.resolve_endpoint(__MODULE__)
```

and `BridgeEx.resolve_endpoint/1` could be something like

```
def resolve_endpoint(nil, mod_name) do
  case mod_name |> Application.get_application() |> Application.get_env(mod_name, []) |> Keyword.get(:endpoint, :undefined) do
  :undefined -> raise "No configured endpoint for bridge #{mod_name}"
  value -> value
end

def resolve_endpoint(endpoint, mod_name), do: endpoint
```

This way we can still use our new `.env` + `runtime.exs` goodies, AND override the call/3 through options, which is a reasonable use case imho.
Is there something wrong in my reasoning? :pray:

**Comment by cpiemontese** [comment]
I understand the reasoning but I fear this would not work in case of multiple dependencies using the bridge which in our case means umbrella apps i.e. Stonehenge.

We have a global configuration for the `log_options` which is now deprecated because in Stonehenge they couldn't use it and instead they needed `log_options` at the actual `use BridgeEx.Graphql` level. If migrated to OTP 25 and they wanted to use runtime endpoints I think they'd need them to be passed at the `call` level or some way other than `config`

Let me know if this makes sense

Another idea: we could define an [overridable function](https://hexdocs.pm/elixir/1.13.4/Kernel.html#defoverridable/1) to get the endpoint which would default to `@endpoint`

**Comment by mbusi** [comment]
I still don't understand why the arg removal from the macro should mean that we NEED an arg in the `call` function (or an `Application.get_env` call); shouldn't the endpoint be an init argument for the bridge? if we move the macro logic to an init funciton, THAT is what should contain the endpoint argument IMO

**Comment by chrooti** [comment]
```elixir
defmodule Test do
  defmacro prova(opts) when is_list(opts) do
    ep = Keyword.get(opts, :endpoint)
    quote do
      def func() do
        unquote(ep) |> IO.inspect
      end
    end
  end
end

defmodule TestDue do
  require Test 
  Test.prova(endpoint: Application.get_env(:test))
end

TestDue.func

```

poc of what I have in mind, I hope it's self explanatory

**Comment by chrooti** [comment]
> Wouldn't an init function require that we used the bridge as a struct rather than as a macro?

that should be solved by what I did in the PoC above (i.e. unquote the single values inside the keyword list instead that the list itself)

**Comment by cpiemontese** [comment]
@chrooti we are trying out the substitution but it fails on something like this

```elixir
defmodule Test do
  defmacro prova(opts) when is_list(opts) do
    ep = Keyword.get(opts, :endpoint)
    quote do
      def func() do
        unquote(ep) |> IO.inspect
      end
    end
  end
end

defmodule TestDue do
  require Test

  a = 1

  Test.prova(endpoint: Application.get_env(:app, :test, "asdf #{a}"))
end

TestDue.func
```

**Comment by chrooti** [comment]
```elixir
Test.prova(endpoint: Application.get_env(:app, :test, "asdf #{a}"))
```

becomes
```elixir
Test.prova(endpoint: Application.get_env(:app, :test, "asdf #{unquote(a)}"))
```

**Comment by cpiemontese** [comment]
You are right, but the caller needing to write `unquote` seems kinda bad tbh

**Comment by mbusi** [comment]
> > I still don't understand why the arg removal from the macro should mean that we NEED an arg in the call function (or an Application.get_env call); shouldn't the endpoint be an init argument for the bridge? if we move the macro logic to an init funciton, THAT is what should contain the endpoint argument IMO
> 
> Wouldn't an init function require that we used the bridge as a struct rather than as a macro? E.g.
> 
> ```elixir
> bridge = BridgeEx.Graphql.init(Application.get_env(:app, :endpoint))
> BridgeEx.Graphql.call(bridge, ...)
> ```

More a process than a struct but yes, exaclty, and I don't see a problem with that; you could even let your clients call you bridge by name so they don't have to store the pid anywhere (with an Agent that register itself with the module name)

Now that I have a bit more context (looking at the code :D) I understand what @zoten meant on slack with 
```
silly question: for the sake of migrating services, couldn't that be that if you don't explicitly pass an :endpoint configuration (so the updated services can simply delete it), under the hood it defaults to Application.get_env(:bridge_ex, :whatever) (without application namespace, or we could also inject an use of Application.get_application and use app specific settings)?
```
in fact it makes sens to me (if you do it at runtime, with a properly explicit error message in case the env var is not found and properly document it in the readme), so you can set any arg either via macro or via env var

i.e, you can use the lib like this:
```elixir
defmodule MyBridgeModule
  use BridgeEx, endpoint: "http://localhost", headers:[{"Content-Type", "application/json}]
end
```
or like this:
```elixir
defmodule MyBridgeModule
  use BridgeEx
end
```
and then set in your config file
```elixir
config :bridge_ex, MyBridgeModule, endpoint: "http://localhost", headers:[{"Content-Type", "application/json}]
```
^ this could go in your `runtime.exs` file like so:
```elixir
config :bridge_ex, MyBridgeModule, endpoint: System.get_env("MY_BRIDGE_ENDPOINT"), headers:[{"Content-Type", "application/json}]
```

Is it what you meant @zoten ? :D

**Comment by mbusi** [comment]
To implement it, you could do something like (dumb 2 mins example):
```elixir
defmacro __using__(opts) when is_list(opts) do
  quote do
    if Keyword.has_key?(unquote(opts), :endpoint) do
      defp get_endpoint() do
        Keyword.fetch!(unquote(opts), :endpoint)
      end
    else
      defp get_endpoint() do
        Application.fetch_env(:bridge_ex, ...) # you could add a raise here
      end
    end
  end
end
```
or in a more naive way, in your `call` function, do something like:
```elixir
endpoint = @endpoint or Application.get_env(:bridge_ex, __MODULE__) or raise "you done fucked up son"
```

Pardon my rusty elixir :pray:

**Comment by mbusi** [comment]
> I understood better @zoten's proposal and I think a per-module config shouldn't cause clashes in umbrella apps.
> 
> I refactored the code quite a bit so that the behaviour now is:
> 
> * if you have any config defined after `use` then the bridge assumes everything is compile time
> * if you don't have any config defined after `use` the bridge assumes everything is "runtime" (i.e. read from config)
> 
> Let me know what you think.

Why not making it a 'per config' thing?

**Comment by cpiemontese** [comment]
@mbusi I didn't want to encourage having split configs, but maybe that's the point of this i.e. having runtime stuff in a runtime file and compile stuff at the bridge level

**Review by zoten — DISMISSED** [review]

**Comment by mbusi** [comment]
> @mbusi I didn't want to encourage having split configs, but maybe that's the point of this i.e. having runtime stuff in a runtime file and compile stuff at the bridge level

yeah exactly, that was my idea :D (but I am ok with anything tbh)

**Comment by cpiemontese** [comment]
I updated the code to make every option independent from each other and configurable separately either at compile time and runtime & I updated the docs & CHANGELOG :muscle:

**Review by mbusi — APPROVED** [review]

File: lib/graphql.ex [hunk 1]
```diff
@@ -36,46 +36,86 @@ defmodule BridgeEx.Graphql do
       endpoint: "http://my-api.com/graphql",
       auth0: [enabled: true, audience: "target-audience"]
   end
+
+  # If you would like to configure this options to use runtime values,
+  # you can do so through your config e.g.
+  config :bridge_ex, MyBridge,
+    endpoint: "http://my-api.com/graphql"
```
Comment (Fs00) [inline lib/graphql.ex:43]:
```suggestion
  # If you would like to configure these options to use values evaluated at runtime,
  # you can do so through your application config, e.g.
  config :bridge_ex, MyBridge,
    endpoint: Application.fetch_env!(:my_app, :service_endpoint)
```

**Review by Fs00 — COMMENTED** [review]
Various wording improvements to the documentation, the rest already looks good 👍

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,7 +5,15 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [Next]
+## [Unreleased]
+
+### Added
+
+- options can be now be configured through `config` for each single bridge
```
Comment (chess4ever) [inline CHANGELOG.md:12]:
```suggestion
- options can now be configured through `config` for each single bridge
```

**Review by Fs00 — DISMISSED** [review]

**Review by Fs00 — APPROVED** [review]

**Pull request merged**